pyccel/ast/headers.py

```diff
@@ -189,29 +189,83 @@ def create_definition(self):
         for args_ in product(*dtypes):
             args = []
             for i, d in enumerate(args_):
-                dtype    = d['datatype']
-                allocatable = d['allocatable']
-                is_pointer = d['is_pointer']
-                precision = d['precision']
-                rank = d['rank']
-
-                order = None
-                shape = None
-                if rank >1:
-                    order = d['order']
-
-                if isinstance(dtype, str):
-                    try:
-                        dtype = datatype(dtype)
-                    except ValueError:
-                        #TODO check if it's a class type before
-                        dtype =  DataTypeFactory(str(dtype), ("_name"))()
-                        is_pointer = True
-                arg_name = 'arg_{0}'.format(str(i))
-                arg = Variable(dtype, arg_name,
-                               allocatable=allocatable, is_pointer=is_pointer,
-                               rank=rank, shape=shape ,order = order, precision = precision,
-                               is_argument=True)
+                if (d['isfunc']):
+                    decs = []
+                    results = []
+                    for dc in d['decs']:
+                        dtype    = dc['datatype']
+                        allocatable = dc['allocatable']
+                        is_pointer = dc['is_pointer']
+                        precision = dc['precision']
+                        rank = dc['rank']
+
+                        order = None
+                        shape = None
+                        if rank >1:
+                            order = dc['order']
+
+                        if isinstance(dtype, str):
+                            try:
+                                dtype = datatype(dtype)
+                            except ValueError:
+                                dtype = DatatypeFactory(str(dtype), ("_name"))()
+                                is_pointer = True
+                        arg = Variable(dtype, '',
+                                    allocatable=allocatable, is_pointer=is_pointer,
+                                    rank=rank, shape=shape ,order = order, precision = precision,
+                                    is_argument=True)
```

Comment:
It looks like a lot of the code here is similar to what is in the else. Could it be extracted to a new function to avoid repetition?